## Compressing a Resnet-50 trained on Imagenet

#### To replicate results with pretrained models please download the following models

1. resnet50_imagenet_pytorch.pth
2. resnet50_imagenet_pytorch_small_cup_t_point_65.pth
3. resnet50_imagenet_pytorch_small_cup_t_point_70.pth
4. resnet50_imagenet_pytorch_lr_point1_cupSS_K_point03_b_point3.pth

In [1]:
import sys; sys.argv=['']; 
sys.path.insert(0, '../')
del sys
%load_ext autoreload
%autoreload 2
import matplotlib.pyplot as plt
%matplotlib inline

import os
os.environ["CUDA_VISIBLE_DEVICES"] = '5,6,7'

import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torch.nn.parallel
import torch.backends.cudnn as cudnn
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.utils.data
import torch.utils.data.distributed
import torch.utils.model_zoo as model_zoo
import torchvision
import torchvision.transforms as transforms
import torchvision.datasets as datasets
from torchvision import datasets, transforms
from tensorboardX import SummaryWriter

import numpy as np
import random
import time
import copy

# os.environ["CUDA_VISIBLE_DEVICES"] = '1'

from src.imagenet_utils import train,validate,save_checkpoint,AverageMeter,ProgressMeter
from src.imagenet_utils import adjust_learning_rate,accuracy,adjust_learning_rate_pytorch_retrain, adjust_learning_rate_pytorch_cupss
from src.utils import plot_tsne,fancy_dendrogram,save_obj,load_obj
from src.model import VGG,load_model
from src.prune_model import prune_model
from src.cluster_model import cluster_model
from src.train_test import adjust_learning_rate_nips,adjust_learning_rate_iccv

### Specify imagenet data path

In [14]:
parser = argparse.ArgumentParser(description='PyTorch ImageNet Training')

### add path to dataset here #####
parser.add_argument('--data', type=str,default='/localscratch/mgh/Rahul_Imagenet/',metavar='S',help='path to dataset')

_StoreAction(option_strings=['--data'], dest='data', nargs=None, const=None, default='/localscratch/mgh/Rahul_Imagenet/', type=<class 'str'>, choices=None, help='path to dataset', metavar='S')

In [3]:

parser.add_argument('--start-epoch', default=0, type=int, metavar='N',
                    help='manual epoch number (useful on restarts)')
parser.add_argument('-b', '--batch-size', default=256, type=int,
                    metavar='N',
                    help='mini-batch size (default: 256), this is the total '
                         'batch size of all GPUs on the current node when '
                         'using Data Parallel or Distributed Data Parallel')
parser.add_argument('-j', '--workers', default=6, type=int, metavar='N',
                    help='number of data loading workers (default: 4)')
parser.add_argument('--epochs', default=90, type=int, metavar='N',
                    help='number of total epochs to run')
parser.add_argument('--lr', '--learning-rate', default=0.01, type=float,
                    metavar='LR', help='initial learning rate')
parser.add_argument('--momentum', default=0.9, type=float, metavar='M',
                    help='momentum')
parser.add_argument('--weight-decay', '--wd', default=1e-4, type=float,
                    metavar='W', help='weight decay (default:$ 1e-4)')
parser.add_argument('-p', '--print-freq', default=500, type=int,
                    metavar='N', help='print frequency (default: 10)')
parser.add_argument('--resume', default='', type=str, metavar='PATH',
                    help='path to latest checkpoint (default: none)')
parser.add_argument('-e', '--evaluate', dest='evaluate', action='store_true',
                    help='evaluate model on validation set')
parser.add_argument('--pretrained', dest='pretrained', action='store_true',
                    help='use pre-trained model')
parser.add_argument('--world-size', default=-1, type=int,
                    help='number of nodes for distributed training')
parser.add_argument('--rank', default=-1, type=int,
                    help='node rank for distributed training')
parser.add_argument('--dist-url', default='tcp://224.66.41.62:23456', type=str,
                    help='url used to set up distributed training')
parser.add_argument('--dist-backend', default='nccl', type=str,
                    help='distributed backend')
parser.add_argument('--no-cuda', action='store_true', default=False,
                    help='disables CUDA training')
parser.add_argument('--seed', type=int, default=12346, metavar='S',
                    help='random seed (default: 12346)')
parser.add_argument('--num_output', type=int, default=10, metavar='S',
                    help='number of classes(default: 10)')
parser.add_argument('--log-interval', type=int, default=100, metavar='N',
                    help='how many batches to wait before logging training status')
parser.add_argument('--checkpoint_path', type=str, default='./checkpoints/resnet50_imagenet_pytorch.pth', metavar='S',
                    help='path to store model training checkpoints')
parser.add_argument('--gpu', type=int, default=0, nargs='+', help='used gpu')
parser.add_argument('--multiprocessing-distributed', action='store_true',
                        help='Use multi-processing distributed training to launch ')#,
#                          'N processes per node, which has N GPUs. This is the ',
#                          'fastest way to use PyTorch for either single node or ',
#                          'multi node data parallel training')

args = parser.parse_args()

use_cuda = not args.no_cuda and torch.cuda.is_available()
# if use_cuda:
#     print('using gpu',args.gpu)
#     os.environ["CUDA_VISIBLE_DEVICES"] = ','.join(str(x) for x in args.gpu)
device = torch.device("cuda" if use_cuda else "cpu")

# Data loading code
traindir = os.path.join(args.data, 'train')
valdir = os.path.join(args.data, 'val')
normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                 std=[0.229, 0.224, 0.225])

train_dataset = datasets.ImageFolder(
    traindir,
    transforms.Compose([
        transforms.RandomResizedCrop(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        normalize,
    ]))

train_sampler = None

train_loader = torch.utils.data.DataLoader(
    train_dataset, batch_size=args.batch_size, shuffle=(train_sampler is None),
    num_workers=args.workers, pin_memory=True, sampler=train_sampler)

val_loader = torch.utils.data.DataLoader(
    datasets.ImageFolder(valdir, transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        normalize,
    ])),
    batch_size=args.batch_size, shuffle=False,
    num_workers=args.workers, pin_memory=True)

criterion = nn.CrossEntropyLoss().cuda(args.gpu)


writer = SummaryWriter('logs/resnet50_imagenet/')

#set all seeds for reproducability
def set_random_seed(seed):    
    random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(args.seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    torch.backends.cudnn.deterministic = True

set_random_seed(args.seed)

torch.cuda.set_device(args.gpu)

### Train baseline model

In [4]:
set_random_seed(args.seed)

args.lr = 0.1

resnet50 = torchvision.models.resnet50(pretrained=False)
optimizer = torch.optim.SGD(resnet50.parameters(), args.lr,
                            momentum=args.momentum,
                            weight_decay=args.weight_decay)

resnet50 = torch.nn.DataParallel(resnet50)
resnet50.cuda(args.gpu)

best_val_acc = 0

if not os.path.isfile(args.checkpoint_path):
    for epoch in range(args.start_epoch, args.epochs):
        adjust_learning_rate_pytorch_retrain(optimizer, epoch, args)        

        # train for one epoch
        train_loss,train_top1,train_top5 = train(train_loader, resnet50, criterion, optimizer, epoch, args)

        # evaluate on validation set
        val_loss,val_top1,val_top5 = validate(val_loader, resnet50, criterion, args)
        
        if val_top1 > best_val_acc:  
            torch.save(resnet50, args.checkpoint_path, pickle_protocol=4)            
            best_val_acc = val_top1    

        writer.add_scalars('resnet50_imagenet_pytorch_schedule/loss',{'train_loss': train_loss,
                                        'val_loss' : val_loss}, epoch)
        writer.add_scalars('resnet50_imagenet_pytorch_schedule/accuracy',{'train_top1': train_top1,
                                                  'val_top1': val_top1,
                                                  'train_top5': train_top5,
                                                  'val_top5': val_top5}, epoch) 
else:   
    resnet50 = torch.load(args.checkpoint_path).module
    _,val_top1,val_top5 = validate(val_loader, resnet50, criterion, args, verbose=False)
    
print('loaded model with top1 : {}, top5 : {}'.format(val_top1,val_top5))

 * Acc@1 75.868 Acc@5 92.868
loaded model with top1 : 75.86799621582031, top5 : 92.86799621582031


### Compressing using CUP

#### Summary

1. T = 0.65 : large top-1 75.87, small top-1 75.07, top-1 drop 0.80, large top-5 92.87, small top-5 92.30, top-5 drop 0.56          
         Number of FLOPs: 8.21G
         Number of FLOPs: 3.76G

2. T = 0.675 : large top-1 75.87, small top-1 74.73, top-1 drop 1.13, large top-5 92.87, small top-5 92.14, top-5 drop 0.72
         
         Number of FLOPs: 8.21G
         Number of FLOPs: 3.53G
         
3. T = 0.70 : large top-1 75.87, small top-1 74.60, top-1 drop 1.27, large top-5 92.87, small top-5 92.06, top-5 drop 0.81

         Number of FLOPs: 8.21G
         Number of FLOPs: 3.33G

4. T = 0.725 : large top-1 75.87, small top-1 74.42, top-1 drop 1.45, large top-5 92.87, small top-5 91.74, top-5 drop 1.13

         Number of FLOPs: 8.21G
         Number of FLOPs: 3.11G
         

#### T = 0.65

#### Load pre-trained model

In [5]:
resnet50 = torch.load(args.checkpoint_path).module
_,val_top1,val_top5 = validate(val_loader, resnet50, criterion, args, verbose=False)
best_val_acc = val_top5
print('loaded model with top1 : {}, top5 : {}'.format(val_top1,val_top5))

 * Acc@1 75.868 Acc@5 92.868
loaded model with top1 : 75.86799621582031, top5 : 92.86799621582031


In [6]:
cluster_args = {
    'cluster_layers' : {4:0,6:0,13:0,15:0,20:0,22:0,27:0,29:0,36:0,38:0,43:0,45:0,50:0,52:0,57:0,59:0,66:0,68:0,73:0,75:0,80:0,82:0,87:0,89:0,94:0,96:0,101:0,103:0,110:0,112:0,117:0,119:0},
    'conv_feature_size' : 1,
    'features' : 'both',
    'channel_reduction' : 'fro',
    'use_bias' : False,
    'reshape_exists' : False,
    'linkage_method' : 'ward',
    'distance_metric' : 'euclidean',
    'cluster_criterion' : 'hierarchical',
    'distance_threshold' : 0.65,
    'merge_criterion' : 'max_l2_norm',
    'verbose' : False
}

path = args.checkpoint_path[:-4] + '_small_cup_t_point_65.pth' 
model_modifier = cluster_model(resnet50,cluster_args)
resnet50_clustered = model_modifier.cluster_model()#[int(nodes*drop_percentage) for nodes in [500,300]])
optimizer = torch.optim.SGD(resnet50_clustered.parameters(), args.lr,
                            momentum=args.momentum,
                            weight_decay=args.weight_decay)

resnet50_clustered = torch.nn.DataParallel(resnet50_clustered)
resnet50_clustered.cuda(args.gpu)

_,top1_acc_no_retrain,top5_acc_no_retrain = validate(val_loader, resnet50_clustered, criterion, args, verbose=False)

best_val_acc = 0

if not os.path.isfile(path):
    for epoch in range(args.start_epoch, args.epochs):
        adjust_learning_rate_pytorch_retrain(optimizer, epoch, args)        

        # train for one epoch
        train_loss,train_top1,train_top5 = train(train_loader, resnet50_clustered, criterion, optimizer, epoch, args)

        # evaluate on validation set
        val_loss,val_top1,val_top5 = validate(val_loader, resnet50_clustered, criterion, args)
        
        if val_top1 > best_val_acc:  
            torch.save(resnet50_clustered, path, pickle_protocol=4)            
            best_val_acc = val_top1    

        writer.add_scalars('resnet50_imagenet_t_point_65_pytorch/loss',{'train_loss': train_loss,
                                        'val_loss' : val_loss}, epoch)
        writer.add_scalars('resnet50_imagenet_t_point_65_pytorch/accuracy',{'train_top1': train_top1,
                                                  'val_top1': val_top1,
                                                  'train_top5': train_top5,
                                                  'val_top5': val_top5}, epoch) 
else:   
    resnet50_clustered = torch.load(path).module
    _,test_top1_acc,test_top5_acc = validate(val_loader, resnet50_clustered, criterion, args, verbose=False)
        
print('large top-1 {:.2f}, small top-1 {:.2f}, top-1 drop {:.2f}, large top-5 {:.2f}, small top-5 {:.2f}, top-5 drop {:.2f}'.format(val_top1,test_top1_acc,val_top1-test_top1_acc,val_top5,test_top5_acc,val_top5-test_top5_acc))       

 * Acc@1 0.156 Acc@5 0.856
 * Acc@1 75.072 Acc@5 92.304
large top-1 75.87, small top-1 75.07, top-1 drop 0.80, large top-5 92.87, small top-5 92.30, top-5 drop 0.56


In [7]:
from src.compute_flops import print_model_param_nums,print_model_param_flops

resnet50_clustered = resnet50_clustered.module if type(resnet50_clustered) == torch.nn.DataParallel else resnet50_clustered

print_model_param_flops(resnet50.cpu(),input_res=224)
print_model_param_flops(resnet50_clustered.cpu(),input_res=224)

  + Number of FLOPs: 8.21G
  + Number of FLOPs: 3.76G


3761212866.0

#### T = 0.70

#### Load pre-trained model

In [8]:
resnet50 = torch.load(args.checkpoint_path).module
_,val_top1,val_top5 = validate(val_loader, resnet50, criterion, args, verbose=False)
best_val_acc = val_top5
print('loaded model with top1 : {}, top5 : {}'.format(val_top1,val_top5))

 * Acc@1 75.868 Acc@5 92.868
loaded model with top1 : 75.86799621582031, top5 : 92.86799621582031


In [9]:
cluster_args = {
    'cluster_layers' : {4:0,6:0,13:0,15:0,20:0,22:0,27:0,29:0,36:0,38:0,43:0,45:0,50:0,52:0,57:0,59:0,66:0,68:0,73:0,75:0,80:0,82:0,87:0,89:0,94:0,96:0,101:0,103:0,110:0,112:0,117:0,119:0},
    'conv_feature_size' : 1,
    'features' : 'both',
    'channel_reduction' : 'fro',
    'use_bias' : False,
    'reshape_exists' : False,
    'linkage_method' : 'ward',
    'distance_metric' : 'euclidean',
    'cluster_criterion' : 'hierarchical',
    'distance_threshold' : 0.70,
    'merge_criterion' : 'max_l2_norm',
    'verbose' : False
}

path = args.checkpoint_path[:-4] + '_small_cup_t_point_70.pth' 
model_modifier = cluster_model(resnet50,cluster_args)
resnet50_clustered = model_modifier.cluster_model()#[int(nodes*drop_percentage) for nodes in [500,300]])
optimizer = torch.optim.SGD(resnet50_clustered.parameters(), args.lr,
                            momentum=args.momentum,
                            weight_decay=args.weight_decay)

resnet50_clustered = torch.nn.DataParallel(resnet50_clustered)
resnet50_clustered.cuda(args.gpu)

_,top1_acc_no_retrain,top5_acc_no_retrain = validate(val_loader, resnet50_clustered, criterion, args, verbose=False)

best_val_acc = 0

if not os.path.isfile(path):
    for epoch in range(args.start_epoch, args.epochs):
        adjust_learning_rate_pytorch_retrain(optimizer, epoch, args)        

        # train for one epoch
        train_loss,train_top1,train_top5 = train(train_loader, resnet50_clustered, criterion, optimizer, epoch, args)

        # evaluate on validation set
        val_loss,val_top1,val_top5 = validate(val_loader, resnet50_clustered, criterion, args)
        
        if val_top1 > best_val_acc:  
            torch.save(resnet50_clustered, path, pickle_protocol=4)            
            best_val_acc = val_top1    

        writer.add_scalars('resnet50_imagenet_t_point_70_pytorch/loss',{'train_loss': train_loss,
                                        'val_loss' : val_loss}, epoch)
        writer.add_scalars('resnet50_imagenet_t_point_70_pytorch/accuracy',{'train_top1': train_top1,
                                                  'val_top1': val_top1,
                                                  'train_top5': train_top5,
                                                  'val_top5': val_top5}, epoch) 
else:   
    resnet50_clustered = torch.load(path).module
    _,test_top1_acc,test_top5_acc = validate(val_loader, resnet50_clustered, criterion, args, verbose=False)
    
print('large top-1 {:.2f}, small top-1 {:.2f}, top-1 drop {:.2f}, large top-5 {:.2f}, small top-5 {:.2f}, top-5 drop {:.2f}'.format(val_top1,test_top1_acc,val_top1-test_top1_acc,val_top5,test_top5_acc,val_top5-test_top5_acc))        

 * Acc@1 0.144 Acc@5 0.730
 * Acc@1 74.596 Acc@5 92.060
large top-1 75.87, small top-1 74.60, top-1 drop 1.27, large top-5 92.87, small top-5 92.06, top-5 drop 0.81


In [10]:
from src.compute_flops import print_model_param_nums,print_model_param_flops

resnet50_clustered = resnet50_clustered.module if type(resnet50_clustered) == torch.nn.DataParallel else resnet50_clustered

print_model_param_flops(resnet50.cpu(),input_res=224)
print_model_param_flops(resnet50_clustered.cpu(),input_res=224)

  + Number of FLOPs: 8.21G
  + Number of FLOPs: 3.33G


3334911739.0

### Try pruning alongside training the model (CUP-SS)

### Summary

1. K=0.03, b = 0.3 :  small top-1 74.40, small top-5 91.99

  + Number of FLOPs: 8.21G
  + Number of FLOPs: 3.74G


2. K=0.03, b = 0.4 : small top-1 74.31, small top-5 92.10

  + Number of FLOPs: 8.21G
  + Number of FLOPs: 3.80G


3. K=0.03, b = 0.5 : small top-1 74.39, small top-5 91.94

  + Number of FLOPs: 8.21G
  + Number of FLOPs: 3.72G



#### K=0.03, b = 0.3

In [11]:
from src.compute_flops import print_model_param_nums,print_model_param_flops

set_random_seed(args.seed)

resnet50 = torchvision.models.resnet50(pretrained=False)
resnet50_clustered = torch.nn.DataParallel(resnet50)
resnet50_clustered.cuda(args.gpu)


cluster_args = {
    'cluster_layers' : {4:0,6:0,13:0,15:0,20:0,22:0,27:0,29:0,36:0,38:0,43:0,45:0,50:0,52:0,57:0,59:0,66:0,68:0,73:0,75:0,80:0,82:0,87:0,89:0,94:0,96:0,101:0,103:0,110:0,112:0,117:0,119:0},
    'conv_feature_size' : 1,
    'features' : 'both',
    'channel_reduction' : 'fro',
    'use_bias' : False,
    'reshape_exists' : False,
    'linkage_method' : 'ward',
    'distance_metric' : 'euclidean',
    'cluster_criterion' : 'hierarchical',
    'distance_threshold' : 0.725,
    'merge_criterion' : 'max_l2_norm',
    'verbose' : False
}

path = args.checkpoint_path[:-4] + '_lr_point1_cupSS_K_point03_b_point3.pth' 

T_values = {}
slope, b = 0.03, 0.3

for epoch in range(1,args.epochs+1):
    T_values[epoch] = slope * (epoch - 1) + b

In [12]:
args.epochs = 90
args.lr = 0.1
best_val_acc = 0
flag = True


if not os.path.isfile(path):
    for epoch in range(1, args.epochs+1):          
        if epoch in T_values.keys() and flag:
            print('changing T value to {}'.format(T_values[epoch]))
            cluster_args['distance_threshold'] = T_values[epoch]        
            model_modifier = cluster_model(resnet50_clustered.module,cluster_args)
            resnet50_clustered = model_modifier.cluster_model()            
            resnet50_clustered = torch.nn.DataParallel(resnet50_clustered)
            resnet50_clustered.cuda(args.gpu)
            
            optimizer = torch.optim.SGD(resnet50_clustered.parameters(), args.lr,
                                        momentum=args.momentum,
                                        weight_decay=args.weight_decay)
            T = T_values[epoch]            
            flops = print_model_param_flops(copy.deepcopy(resnet50_clustered.module).cpu(),input_res=224)
            if (flops/1e9 <= 3.81):
                print('stop filter pruning')
                flag = False
                
                
        adjust_learning_rate_pytorch_retrain(optimizer, epoch, args)        

        # train for one epoch
        train_loss,train_top1,train_top5 = train(train_loader, resnet50_clustered, criterion, optimizer, epoch, args)

#         evaluate on validation set
        val_loss,val_top1,val_top5 = validate(val_loader, resnet50_clustered, criterion, args)
        
        if val_top1 > best_val_acc:  
            torch.save(resnet50_clustered, path, pickle_protocol=4)            
            best_val_acc = val_top1    

        writer.add_scalars('resnet50_imagenet_lr_point1_k_point_03_b_point3/loss',{'train_loss': train_loss,
                                        'val_loss' : val_loss}, epoch)
        writer.add_scalars('resnet50_imagenet_lr_point1_k_point_03_b_point3/accuracy',{'train_top1': train_top1,
                                                  'val_top1': val_top1,
                                                  'train_top5': train_top5,
                                                  'val_top5': val_top5}, epoch) 
       
        writer.add_scalars('resnet50_imagenet_lr_point1_k_point_03_b_point3/flops',{'flops': flops}, epoch) 
        writer.add_scalars('resnet50_imagenet_lr_point1_k_point_03_b_point3/T',{'T': T}, epoch) 
else:
    resnet50_clustered = torch.load(path).module
    val_loss,val_top1,val_top5 = validate(val_loader, resnet50_clustered, criterion, args, verbose=False)
        
print('small top-1 {:.2f}, small top-5 {:.2f}'.format(val_top1,val_top5))        

 * Acc@1 74.400 Acc@5 91.986
small top-1 74.40, small top-5 91.99


In [13]:
from src.compute_flops import print_model_param_nums,print_model_param_flops

resnet50_clustered = resnet50_clustered.module if type(resnet50_clustered) == torch.nn.DataParallel else resnet50_clustered

print_model_param_flops(resnet50.cpu(),input_res=224)
print_model_param_flops(resnet50_clustered.cpu(),input_res=224)

  + Number of FLOPs: 8.21G
  + Number of FLOPs: 3.74G


3744130339.0